plugin/code_actions.py

```diff
@@ -1,60 +1,74 @@
 from .core.registry import sessions_for_view, client_from_session
+from .core.registry import windows
 from .core.settings import settings
 from .core.typing import Any, List, Dict, Callable, Optional, Union, Tuple, Mapping, TypedDict
 from .core.url import filename_to_uri
-from .core.views import region_to_range
+from .core.views import entire_content_range, region_to_range
 from .diagnostics import filter_by_point, view_diagnostics
+import sublime
+import sublime_plugin
 
 CodeActionOrCommand = TypedDict('CodeActionOrCommand', {
     'title': str,
     'command': Union[dict, str],
-    'edit': dict
+    'edit': dict,
+    'kind': Optional[str]
 }, total=False)
 CodeActionsResponse = Optional[List[CodeActionOrCommand]]
 CodeActionsByConfigName = Dict[str, List[CodeActionOrCommand]]
 
 
-class CodeActionsAtLocation(object):
-
-    def __init__(self, on_complete_handler: Callable[[CodeActionsByConfigName], None]) -> None:
+class CodeActionsCollector(object):
+    def __init__(self, on_complete_handler: Callable[[CodeActionsByConfigName], None]):
+        self._on_complete_handler = on_complete_handler
         self._commands_by_config = {}  # type: CodeActionsByConfigName
         self._requested_configs = []  # type: List[str]
-        self._on_complete_handler = on_complete_handler
+        self._all_requested = False
 
-    def collect(self, config_name: str) -> Callable[[CodeActionsResponse], None]:
+    def __enter__(self) -> 'CodeActionsCollector':
+        return self
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        self._all_requested = True
+        self._notify_if_all_finished()
+
+    def create_collector(self, config_name: str) -> Callable[[CodeActionsResponse], None]:
         self._requested_configs.append(config_name)
-        return lambda actions: self.store(config_name, actions)
+        return lambda actions: self._collect_response(config_name, actions)
 
-    def store(self, config_name: str, actions: CodeActionsResponse) -> None:
+    def _collect_response(self, config_name: str, actions: CodeActionsResponse) -> None:
         self._commands_by_config[config_name] = actions or []
-        if len(self._requested_configs) == len(self._commands_by_config):
-            self._on_complete_handler(self._commands_by_config)
+        self._notify_if_all_finished()
 
-    def deliver(self, recipient_handler: Callable[[CodeActionsByConfigName], None]) -> None:
-        recipient_handler(self._commands_by_config)
+    def _notify_if_all_finished(self) -> None:
+        if self._all_requested and len(self._requested_configs) == len(self._commands_by_config):
+            # Call back on the main thread
+            sublime.set_timeout(lambda: self._on_complete_handler(self._commands_by_config))
+
+    def get_actions(self) -> CodeActionsByConfigName:
+        return self._commands_by_config
 
 
 class CodeActionsManager(object):
     """ Collects and caches code actions"""
```

Comment:
The description here is strange, because there's already a "CodeActionsCollector" class above.